crates/local-cluster-runner/src/node/mod.rs

```diff
@@ -0,0 +1,740 @@
+                self.base_config.common.metadata_store_client.metadata_store_client = MetadataStoreClient::Embedded { address: AdvertisedAddress::Uds(metadata_socket) }
+            }
+
+            pub fn with_random_ports(self) {
+                self.base_config.admin.bind_address =
+                    random_socket_address().expect("to find a random port for the admin server");
+                self.base_config.admin.query_engine.pgsql_bind_address =
+                    random_socket_address().expect("to find a random port for the pgsql server");
+                self.base_config.ingress.bind_address =
+                    random_socket_address().expect("to find a random port for the ingress server");
+            }
+
+            pub fn with_node_name(self, node_name: impl Into<String>) {
+                self.base_config.common.set_node_name(node_name.into());
+            }
+
+            pub fn with_node_id(self, node_id: PlainNodeId) {
+                self.base_config.common.force_node_id = Some(node_id)
+            }
+
+            pub fn with_roles(self, roles: EnumSet<Role>) {
+                self.base_config.common.roles = roles
+            }
+        ))]
+    base_config: Configuration,
+    binary_source: BinarySource,
+    #[builder(default)]
+    args: Vec<String>,
+    #[builder(default = true)]
+    inherit_env: bool,
+    #[builder(default)]
+    env: Vec<(String, String)>,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum NodeStartError {
+    #[error("Failed to absolutize node base path: {0}")]
+    Absolute(io::Error),
+    #[error(transparent)]
+    BinarySourceError(#[from] BinarySourceError),
+    #[error("Failed to create node base directory: {0}")]
+    CreateDirectory(io::Error),
+    #[error("Failed to create or truncate node config file: {0}")]
+    CreateConfig(io::Error),
+    #[error("Failed to create or append to node log file: {0}")]
+    CreateLog(io::Error),
+    #[error("Failed to dump config to bytes: {0}")]
+    DumpConfig(GenericError),
+    #[error("Failed to spawn restate-server: {0}")]
+    SpawnError(io::Error),
+}
+
+impl Node {
+    pub fn node_name(&self) -> &str {
+        self.base_config.node_name()
+    }
+
+    pub fn config(&self) -> &Configuration {
+        &self.base_config
+    }
+
+    pub fn config_mut(&mut self) -> &mut Configuration {
+        &mut self.base_config
+    }
+
+    pub fn new_test_node(
+        node_name: impl Into<String>,
+        base_config: Configuration,
+        binary_source: BinarySource,
+        roles: EnumSet<Role>,
+    ) -> Self {
+        Self::builder()
+            .binary_source(binary_source)
+            .base_config(base_config)
+            .with_node_name(node_name)
+            .with_node_socket()
+            .with_socket_metadata()
+            .with_random_ports()
+            .with_roles(roles)
+            .build()
+    }
+
+    pub fn new_test_nodes_with_metadata(
```

Comment:
I'd appreciate if the public API is documented. For instance, one might wonder how`roles` here apply to nodes. etc.